File: LeetCode/search_in_rotated_sorted_array.cpp
class Solution {
public:
    int bn(vector<int>& nums, int l, int r, int target) {
        if(l > r)
            return -1;
        
        int mid = (l + r) / 2;
        if(nums[mid] == target)
            return mid;
        
        if(nums[l] <= nums[mid]) {
            if(target >= nums[l] && target <= nums[mid])
                return bn(nums, l, mid-1, target);
            return bn(nums, mid+1, r, target);
        }
        
        if(target >= nums[mid] && target <= nums[r])
            return bn(nums, mid+1, r, target);
        return bn(nums, l, mid-1, target);
    }
    
    int search(vector<int>& nums, int target) {
        return bn(nums, 0, nums.size()-1, target);
    }
};

File: LeetCode/search_in_rotated_sorted_array.go
func search(nums []int, target int) int {
    l, r := 0, len(nums)-1

    for l <= r {
        mid := (l + r) / 2

        if nums[mid] == target {
            return mid
        }

        if nums[mid] > nums[r] {
            if nums[mid] > target && target >= nums[l] {
                r = mid - 1
            } else {
                l = mid + 1
            }
        } else if nums[mid] < nums[l] {
            if nums[mid] < target && target <= nums[r] {
                l = mid + 1
            } else {
                r = mid - 1
            }
        } else {
            if nums[mid] < target {
                l = mid + 1
            } else {
                r = mid - 1
            }
        }
    }

    return -1
}
